/*

***********************************************************************************
* Copyright (C) 2018 - 2020, BlockSettle AB
* Distributed under the GNU Affero General Public License (AGPL v3)
* See LICENSE or http://www.gnu.org/licenses/agpl.html
*
**********************************************************************************

*/
syntax = "proto3";

// Messages common between systems (e.g. terminal and PB)
package BlockSettle.Common;


message Empty {}

message NewKeyCompare
{
   string old_key = 1;
   string new_key = 2;
   string server_id = 3;
}


message ArmoryMessage
{
   message Loading
   {
      int32 network_type = 1;
   }

   message Settings
   {
      int32  socket_type = 1;
      int32  network_type = 2;
      string host = 3;
      string port = 4;
      string bip15x_key = 5;
      bool   run_locally = 6;
      string data_dir = 7;
      string executable_path = 8;
      string bitcoin_dir = 9;
      string db_dir = 10;
   }

   message State
   {
      int32  state = 1;
      uint32 top_block = 2;
   }

   message RegisterWallet
   {
      string wallet_id = 1;
      bool   as_new = 2;
      repeated bytes addresses = 3;
   }

   message WalletRegistered
   {
      string wallet_id = 1;
      bool   success = 2;
   }

   message Refresh
   {
      repeated string ids = 1;
      bool online = 2;
   }

   message NewBlock
   {
      uint32 top_block = 1;
      uint32 branch_height = 2;
   }

   message ZCInvalidated
   {
      repeated bytes tx_hashes = 1;
   }

   message TXEntry
   {
      bytes  tx_hash = 1;
      repeated string wallets_id = 2;
      int64  value = 3;
      uint32 block_num = 4;
      uint32 tx_time = 5;
      bool   RBF = 6;
      bool   chained = 7;
      repeated string addresses = 8;
   }

   message ZCReceived
   {
      string request_id = 1;
      repeated TXEntry tx_entries = 2;
   }

   message TXPushData
   {
      bytes expected_tx_hash = 1;
      bytes tx = 2;
   }

   message TXPushRequest
   {
      repeated TXPushData txs_to_push = 1;
      string push_id = 2;	// optional
   }

   enum PushTxResult
   {
      PushTxSuccess = 0;
      PushTxInvalidated = 1;
      PushTxMempoolConflict = 2;
      PushTxMempoolFull = 3;
      PushTxOtherError = 5;
      PushTxAlreadyInChain = 6;
      PushTxAlreadyInMempool = 7;
   }

   message TXPushResult
   {
      string request_id = 1;
      string push_id = 2;
      bytes  tx_hash = 3;
      PushTxResult result = 4;
      string error_message = 5;
   }

   message WalletUnconfirmedTarget
   {
      string wallet_id = 1;
      uint32 conf_count = 2;
   }

   message WalletIDs
   {
      repeated string wallet_ids = 1;
   }

   message AddressTxNsResponse
   {
      message AddressTxNs {
         string wallet_id = 1;

         message AddressTxN {
            bytes  address = 1;
            uint32 txn = 2;
         }
         repeated AddressTxN txns = 2;
      }
      repeated AddressTxNs wallet_txns = 1;
   }

   message WalletBalanceResponse
   {
      message WalletBalance {
         string wallet_id = 1;
         uint64 full_balance = 2;
         uint64 spendable_balance = 3;
         uint64 unconfirmed_balance = 4;
         uint32 address_count = 5;

         message AddressBalance {
            bytes  address = 1;
            uint64 full_balance = 2;
            uint64 spendable_balance = 3;
            uint64 unconfirmed_balance = 4;
         }
         repeated AddressBalance addr_balances = 6;
      }
      repeated WalletBalance balances = 1;
   }

   message TXHashes
   {
      repeated bytes tx_hashes = 1;
      bool disable_cache = 2;
   }

   message Transactions
   {
      repeated bytes transactions = 1;
   }

   oneof data
   {
      Loading          loading = 1;
      Empty            ready   = 2;
      Empty            reconnect = 3;
      Empty            settings_request = 4;
      Settings         settings_response = 5;
      State            state_changed = 6;
      RegisterWallet   register_wallet = 10;
      WalletRegistered wallet_registered = 11;
      Refresh          refresh = 12;
      NewBlock         new_block = 13;
      ZCInvalidated    zc_invalidated = 14;
      ZCReceived       zc_received = 15;
      TXPushRequest    tx_push = 16;
      TXPushResult     tx_push_result = 17;
      string           tx_push_timeout = 18;
      WalletUnconfirmedTarget set_unconf_target = 19;
      string           unconf_target_set = 20;
      WalletIDs        addr_txn_request = 21;
      AddressTxNsResponse addr_txn_response = 22;
      WalletIDs        wallet_balance_request = 23;
      WalletBalanceResponse wallet_balance_response = 24;
      TXHashes         get_txs_by_hash = 25;
      Transactions     transactions = 26;
   }
}


message ErrorResult
{
   int32  code = 1;
   string text = 2;
}


message WalletSyncInfo
{
   int32  format = 1;
   string id = 2;
   string name = 3;
   string description = 4;
   int32  network_type = 5;
   bool   watch_only = 6;
   repeated int32 encryption_types = 7;
   repeated bytes encryption_keys = 8;

   message KeyRank {
      uint32 m = 1;
      uint32 n = 2;
   }
   KeyRank encryption_rank = 9;
}


message SignerMessage
{
   message WalletsInfo
   {
      repeated WalletSyncInfo wallets = 1;
   }

   message SyncAddresses
   {
      string wallet_id = 1;
      repeated bytes addresses = 2;
   }

   message SyncAddrResult
   {
      string wallet_id = 1;
      int32  status = 2;
   }

   oneof data
   {
      NewKeyCompare  new_key_request  = 1;
      bool           new_key_response = 2;
      ErrorResult    error = 3;
      string         auth_leaf_added = 4;
      Empty          wallets_list_updated = 5;
      Empty          wallet_storage_decrypted = 6;
      Empty          ready = 7;
      Empty          need_new_wallet_prompt = 8;
      Empty          wallets_ready_to_sync = 9;
      bool           window_visible_changed = 10;
      Empty          start_wallets_sync = 11;
      WalletsInfo    wallets_info = 12;
      SyncAddresses  sync_addresses = 13;
      SyncAddrResult sync_addr_result = 14;
   }
}


message WalletsMessage
{
   message Loading
   {
      uint32 nb_wallets = 1;
      repeated string wallet_id = 2;
   }

   message WalletLoaded
   {
      string wallet_id = 1;
   }

   message WalletError
   {
      string wallet_id = 1;
      string error_message = 2;
      int32  error_code = 3;
   }

   message HDWalletData
   {
      message Leaf {
         string id = 1;
         string path = 2;
         bool   ext_only = 3;
         bytes  extra_data = 4;	// optional
      }
      message Group {
         int32 type = 1;
         repeated Leaf leaves = 2;
         bool  ext_only = 3;
         bytes salt = 4;	// optional - for Auth group
      }
      string wallet_id = 1;
      repeated Group groups = 2;
   }

   message WalletData
   {
      message AddressData {
         string index = 1;
         string address = 2;
         string comment = 3;
      }
      message TXComment {
         bytes  tx_hash = 1;
         string comment = 2;
      }

      uint32 high_ext_index = 1;
      uint32 high_int_index = 2;
      repeated AddressData used_addresses = 3;
      repeated AddressData reserved_addresses = 4;
      repeated TXComment   tx_comments = 5;
   }

   oneof data
   {
      Loading      loading = 1;
      Empty        ready   = 2;
      WalletError  error = 3;
      WalletLoaded wallet_loaded = 4;
      WalletLoaded auth_leaf_created = 5;
      WalletLoaded balance_updated = 6;
      WalletLoaded wallet_changed = 7;
      WalletLoaded wallet_meta_changed = 8;
      WalletLoaded wallet_ready = 9;
      WalletLoaded wallet_reset = 10;    
      Empty        start_wallets_sync = 11;
      string       hd_wallet_get = 12;
      HDWalletData hd_wallet = 13;
      string       wallet_get = 14;
      WalletData   wallet_data = 15;
   }
}
